File: Algorithm/Python/914. X of a Kind in a Deck of Cards.py
# -*- coding: utf-8 -*-
# Created on Tue Oct 02 2018 17:7:35
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# Greatest Common Divisor
from collections import Counter
class Solution(object):
    def hasGroupsSizeX(self, deck):
        """
        :type deck: List[int]
        :rtype: bool
        """
        def gcd(x, y):
            while y:
                x, y = y, x%y
            return x
        count = Counter(deck)
        return reduce(gcd, count.values())>1

File: Algorithm/C++/914. X of a Kind in a Deck of Cards.cpp
/*
 * Created on Wed Oct 03 2018 10:48:50
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// greatest commom divisor
class Solution {
public:
    bool hasGroupsSizeX(vector<int>& deck) {
        unordered_map<int, int> count;
        for(int num:deck) count[num]++;
        int result = 0;
        for(auto iterator : count) result = __gcd(iterator.second, result);
        return result>1;
    }
};